### ISSUE TITLE
Feed a pre-trained embedding using a CSV file

### ISSUE THREAD

**Issue by krayush07** [tflearn/tflearn#186]
I have a pre-trained csv file which contains word embeddings for each word in the vocab.

For example, word vector of dimension 3, vocab size, say 150:

food 0.4 -0.2 0.04
is 0.97 1.23 -2.3
caught 1.45 -2.34 0.23
...

I wish to use a LSTM framework in which I need to feed the word vectors from the csv file at the input layer and do not wish to re-train the word embeddings (as in EMBEDDING layer given in example at: https://github.com/tflearn/tflearn/blob/master/examples/nlp/dynamic_lstm.py). I did a lot of search and could not find a simple way to achieve this case. Can someone please guide for this issue?

**Comment by aymericdamien** [comment]
You have different ways to do that:
- One way could be to use Preprocessing class to add a lookup and preprocess your data prior to feeding them to the neural net:
see https://github.com/tflearn/tflearn/blob/master/tflearn/data_preprocessing.py#L100 to add your custom preprocessing method
- Otherwise you can directly use `tf.nn.embedding_lookup`, but it requires you to transform all your words to integer ids, change your sequence of words to sequence of ids, and finally use that tensorflow ops to do the lookup. You can then convert your lookup matrix by applying `W = tf.constant(YOUR_EMBEDDING)`.

**Comment by krayush07** [comment]
Okay, so if ids of word sequence in a sentence be [9, 12, 1, 55, 278, 12], in other sentence be [34, 56, 21] and so on and I need to create a input matrix for these sentences where each word vector is of dimension say 3 and vocab size is 150.

So in tf.nn.embedding_lookup(params, ids, partition_strategy, name, validate_indices)

_ids_ will be array [ [], [], [], ..., [9, 12, 1, 55, 278, 12], [], [34, 56, 21], ... [] ]
_partition strategy_ would be div since the lookup is to be done on sequential word ids?

What would be 'params'? 
Is that a single word vector matrix of dim [vocab size, vector_size] where row1 represents the word vector of word1, row2 represents the word vector of word2 and so on?


What would be the returned matrix? Is that returned matrix is _YOUR_EMBEDDING_ ?

**Comment by aymericdamien** [comment]
Your embedding matrix will looks like this (based on your example):
```
[[0.0    0.0   0.0  ]  <--- needed for consistency (but it will not be used, as id '0' will be masked)
 [0.4   -0.2   0.04]  <--- food
 [0.97 1.23  -2.3 ]  <--- is
 [1.45 -2.34 0.23]  <--- caught
 [ ........               ]  <--- A last id for the unknown words
```
Where ids are defined as follow:
```
food -> 1
is -> 2
caught -> 3
```
Then you can used id 0 as a masking value (to be used with dynamical RNNs, see the example).
Then the sentence 'food is really caught' will be generated as [1 2 4 3].

'params' is your matrix (above) and 'ids' is your incoming tensor.

**Comment by krayush07** [comment]
Thanks, I can now feed the embedding matrix using the placeholders (see codes below). However, I am facing issues while putting a variable size integer matrix (representing the sequence of words in different sentences) in the feed_dict. Infact I am also getting an error on using a fixed size id_array through placeholder.

```
import tflearn
import tensorflow as tf
import numpy as np

vocab_size = 5
embedding_dim = 3

embedding = np.array([[0.0, 0.0, 0.0 ], [0.4, -0.2, .04], [0.97, 1.23 , -2.3], [1.45, -2.34, 0.23], [-1.0, -1.0, -1.0]], np.float32)
idarr =  np.array([[1,2,4,3], [2,3,4], [2,1]], np.int32) #id representing sequence of words in the sentences

W = tf.Variable(tf.constant(0.0, shape=[vocab_size, embedding_dim]), trainable=False, name="W")
embedding_placeholder = tf.placeholder(tf.float32, [vocab_size, embedding_dim])
embedding_init = W.assign(embedding_placeholder)

'''
Wid = tf.Variable(tf.constant(0.0, shape=[2, 4]), trainable=False, name="Wid")
id_placeholder = tf.placeholder(tf.float32, [2, 4])
id_init = Wid.assign(id_placeholder)
'''

trainX = tf.nn.embedding_lookup(embedding_init, id_init, partition_strategy='div', name=None, validate_indices=True)  #not working

with tf.Session() as sess:
	emb_arr, trainX = sess.run([embedding_init, trainX], feed_dict={embedding_placeholder: embedding})  #correctly working when hard-coded `fixed dimension id_array` ([1,2,4,3], [2,3,2,1]..) is put
	#emb_arr, trainX = sess.run([embedding_init, trainX], feed_dict={embedding_placeholder: embedding, id_placeholder: idarr})  #Error with this code
	print(emb_arr)
	print(trainX)
```

**Comment by aymericdamien** [comment]
Tensorflow doesn't accept inconsistent arrays, you have to pad your sequences with 0s at the end. You can set 'dynamic=True' to your RNN layer, and it will perform dynamic calculation.
About your error, try to cast your matrix 'idarr' to float32.

**Comment by krayush07** [comment]
Thanks for the help. It worked.
